### ISSUE TITLE
ENH: Exclude same input geometry from output of nearest_all

### ISSUE THREAD

**Issue by brendan-ward** [pygeos/pygeos#324]
In [this issue](https://github.com/Toblerity/Shapely/issues/1106) in Shapely, there is a need to find the nearest items in a tree constructed of the same geometries that are being used to query the tree, but return the non-self nearest neighbor(s).

One way to approach this would be to provide a new function that does not take new geometry inputs, but instead uses the underlying tree geometries in such a way that we can compare addresses of a geometry to the results returned within the nearest callback function, so as to avoid a more expensive test for equality.  We would then return an arbitrarily large distance for the self-pair, to force the tree to look for the next nearest neighbor.

/cc @liunx7594

**Comment by sgillies** [comment]
@brendan-ward that's a good insight about the cost of an equality test. In https://github.com/Toblerity/Shapely/pull/1166/files#diff-6108807fae8f58167beefb19a5c3270a442c26cc1e5b3b62e0ce25ebbe4673c3R254 I'm promising an `exclusive` keyword argument to make the set of possible results and the input exclusive. One thing that's not great about using the GEOS distance is that all shapes that contain a shape have the same distance: 0.

**Comment by brendan-ward** [comment]
> One thing that's not great about using the GEOS distance is that all shapes that contain a shape have the same distance: 0

This is partly why we return all equidistant results via `nearest_all`, though that does present a bit of an awkward API ("nearest" implies a singular result).  We leave it up to the caller to determine how to then ordinate the results where distance is 0; perhaps they could do so by comparing the distance between centroids or some other metric to better refine that (but it all seems very use specific so I don't know that we can do more here).

It looks like your implementation would extend nicely to `nearest_all` here as something broadcastable to the shape of the inputs, but I think we'd want to work out whether or not we want to express this in terms of avoiding self-nearest neighbors (proposed above, using pointer addresses), avoiding input nearest neighbors (via `exclusive`, using equality test), or both.  I didn't do a performance test on the equality test, so the gains of using the pointer approach may or may not be negligible; it just seemed like it would involve fewer operations.

**Comment by jorisvandenbossche** [comment]
> I didn't do a performance test on the equality test, so the gains of using the pointer approach may or may not be negligible; it just seemed like it would involve fewer operations.

Performance might be a reason to choose the pointer method, but the two methods also can give a different result (there can be multiple geometries in the tree that are not identical, but are spatially equal: the one method only excludes the identical geometry, the other method excludes all equal geometries). Although I don't really have a good sense which of the two behaviours would be most preferable (probably also use case dependent)

**Comment by caspervdw** [comment]
I agree that we should stick to the “equals” and not go for the identity (pointer equality).

I don’t see a pygeos geometry as something that has an identity. They are like integers of strings: immutable and only their value gives them meaning.

Within the context of a tree, we could only do an exclusion by identity if one would specify the query geometry with an index into the tree. (e.g. tree.query(geom_idx) ) but something like that will have a very limited usability.
